pkg/webhook/admission/http_test.go

```diff
@@ -0,0 +1,191 @@
+`)
+		It("should return a response with an error", func() {
+			wh.ServeHTTP(w, req)
+			Expect(w.response).NotTo(BeNil())
+			Expect(w.response).To(Equal(expected))
+		})
+	})
+
+	Describe("can't decode body", func() {
+		req := &http.Request{
+			Header: http.Header{"Content-Type": []string{"application/json"}},
+			Body:   nopCloser{Reader: bytes.NewBufferString("{")},
+		}
+		wh := &Webhook{
+			Type:     types.WebhookTypeMutating,
+			Handlers: []Handler{},
+		}
+		w := &fakeResponseWriter{}
+		expected := []byte(
+			`{"response":{"uid":"","allowed":false,"status":{"metadata":{},"message":"couldn't get version/kind; json parse error: unexpected end of JSON input","code":400}}}
+`)
+		It("should return a response with an error", func() {
+			wh.ServeHTTP(w, req)
+			Expect(w.response).NotTo(BeNil())
+			Expect(w.response).To(Equal(expected))
+		})
+	})
+
+	Describe("empty body after decoding", func() {
+		req := &http.Request{
+			Header: http.Header{"Content-Type": []string{"application/json"}},
+			Body:   nopCloser{Reader: bytes.NewBuffer(nil)},
+		}
+		wh := &Webhook{
+			Type:     types.WebhookTypeMutating,
+			Handlers: []Handler{},
+		}
+		w := &fakeResponseWriter{}
+		expected := []byte(`{"response":{"uid":"","allowed":false,"status":{"metadata":{},"message":"got an empty AdmissionRequest","code":400}}}
+`)
+		It("should return a response with an error", func() {
+			wh.ServeHTTP(w, req)
+			Expect(w.response).NotTo(BeNil())
+			Expect(w.response).To(Equal(expected))
+		})
+	})
+
+	Describe("no webhook type", func() {
+		req := &http.Request{
+			Header: http.Header{"Content-Type": []string{"application/json"}},
+			Body:   nopCloser{Reader: bytes.NewBufferString(`{"request":{}}`)},
+		}
+		wh := &Webhook{
+			Handlers: []Handler{},
+		}
+		w := &fakeResponseWriter{}
+		expected := []byte(`{"response":{"uid":"","allowed":false,"status":{"metadata":{},"message":"you must specify your webhook type","code":400}}}
+`)
+		It("should return a response with an error", func() {
+			wh.ServeHTTP(w, req)
+			Expect(w.response).NotTo(BeNil())
+			Expect(w.response).To(Equal(expected))
+		})
+	})
+
+	Describe("handler can be invoked", func() {
+		req := &http.Request{
+			Header: http.Header{"Content-Type": []string{"application/json"}},
+			Body:   nopCloser{Reader: bytes.NewBufferString(`{"request":{}}`)},
+		}
+		h := &fakeHandler{}
+		wh := &Webhook{
+			Type:     types.WebhookTypeValidating,
+			Handlers: []Handler{h},
+			KVMap:    map[string]interface{}{"foo": "bar"},
+		}
+		w := &fakeResponseWriter{}
+		expected := []byte(`{"response":{"uid":"","allowed":true}}
+`)
+		It("should return a response successfully", func() {
+			wh.ServeHTTP(w, req)
+			Expect(w.response).NotTo(BeNil())
+			Expect(w.response).To(Equal(expected))
+			Expect(h.invoked).To(BeTrue())
+			Expect(h.valueFromContext).To(Equal("bar"))
+		})
+	})
+})
+
+type fakeResponseWriter struct {
```

Comment:
Did you look at `httptest` pkg https://golang.org/pkg/net/http/httptest/ It provides some nice utilities for testing http handlers without running the server.